mlx_lm/generate.py

```diff
@@ -284,21 +285,50 @@ def maybe_quantize_kv_cache(prompt_cache, quantized_kv_start, kv_group_size, kv_
                 )
 
 
+class SearchStrategy(ABC):
+    """
+    A strategy for obtaining a sequence of tokens from the model.
+    """
+
+    @abstractmethod
+    def generate(
+        self,
+        y: mx.array,
+        prompt_cache: List[Any],
+        quantize_cache_fn: Callable[[Any], None],
+        total_prompt_tokens: int,
+    ) -> Generator[Tuple[int, mx.array], None, None]:
+        """
+        Generates a sequence of tokens from the model.
+
+        Args:
+            y (mx.array): The next token sequence.
+            prompt_cache (List[Any]): The prompt cache.
+            quantize_cache_fn (Callable[[Any], None]): A function to quantize the cache.
+            total_prompt_tokens (int): The total number of tokens in the prompt.
+
+        Yields:
+            Tuple[int, mx.array]: A token and a vector of log probabilities.
+        """
+        pass
+
+
 def generate_step(
     prompt: mx.array,
     model: nn.Module,
     *,
     max_tokens: int = 256,
-    sampler: Optional[Callable[mx.array, mx.array]] = None,
+    sampler: Optional[Callable[[mx.array], mx.array]] = None,
```

Comment:
Not related to the main change, but I think the typing was wrong here. The sampler has a single `mx.array` argument and returns an `mx.array`.